docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,305 @@
+# Meaning of `?` variables in K
+
+## Summary
+
+The frontend syntax uses question mark variables of the form `?X` to refer to
+existential variables, and uses `ensures` to specify logical constraints on
+those variables.
+These variables are only allowed to appear in the RHS of a rule.
+
+When K rules are translated to Kore axioms, the `?` variables are existentially
+quantified at the top of the RHS and the `ensures` clause is and-ed to the RHS.
```

Comment:
"In addition, the regular `non-?` variables are universally quantified at the top of the entire rule. When there are both `?`-variables and the regular variables, we quantify regular variables first, followed by `?`-variables.